File: jobs/2026-05-27__14-51-17/extract_with_citations_py__iziMDWq/artifacts/user/llamacloud-task/extract_invoice.py
#!/usr/bin/env python3
"""
Extract structured invoice data with citations using LlamaCloud Extract v2.
"""

import os
import sys
import time
import json
from pathlib import Path

from llama_cloud import LlamaCloud
from pydantic import BaseModel, Field


def main():
    # Read run-id from environment variable
    run_id = os.environ.get("ZEALT_RUN_ID")
    if not run_id:
        print("Error: ZEALT_RUN_ID environment variable is not set", file=sys.stderr)
        sys.exit(1)

    # Define the schema for invoice extraction
    class InvoiceSchema(BaseModel):
        company_name: str = Field(description="The name of the vendor or company issuing the invoice")
        invoice_number: str = Field(description="The invoice number or reference identifier")
        total_amount: float = Field(description="The total amount due on the invoice")

    # Initialize LlamaCloud client (reads LLAMA_CLOUD_API_KEY from environment)
    client = LlamaCloud()

    # Upload the invoice file
    invoice_path = "/home/user/llamacloud-task/sample_invoice.txt"
    external_file_id = f"invoice-{run_id}.txt"

    print(f"Checking for existing file with external_file_id: {external_file_id}")

    # Try to find an existing file with this external_id
    existing_files = list(client.files.list(external_file_id=external_file_id))
    file_id = None

    if existing_files and len(existing_files) > 0:
        file_id = existing_files[0].id
        print(f"Found existing file. File ID: {file_id}")
    else:
        # If no existing file found, upload a new one
        print(f"Uploading new file with external_file_id: {external_file_id}")
        file_response = client.files.create(
            file=invoice_path,
            purpose="extract",
            external_file_id=external_file_id,
        )
        file_id = file_response.id
        print(f"File uploaded successfully. File ID: {file_id}")

    # Create extraction job with citation enabled
    schema_json = InvoiceSchema.model_json_schema()

    extraction_config = {
        "extraction_target": "per_doc",
        "tier": "agentic",
        "cite_sources": True,
    }

    print("Creating extraction job...")
    job = client.extract.create(
        file_input=file_id,
        schema=schema_json,
        configuration=extraction_config,
    )

    job_id = job.id
    print(f"Extract job: {job_id}")

    # Poll for job completion
    print("Polling for job completion...")
    max_attempts = 300  # 5 minutes with 1-second intervals
    attempt = 0

    while attempt < max_attempts:
        job_status = client.extract.get(job_id)
        status = job_status.status

        print(f"Current status: {status}")

        if status in ("COMPLETED", "FAILED", "CANCELLED"):
            break

        time.sleep(1)
        attempt += 1

    status = job_status.status
    print(f"Status: {status}")

    # Check for terminal failure states
    if status in ("FAILED", "CANCELLED"):
        print(f"Error: Job ended with status {status}", file=sys.stderr)
        sys.exit(1)

    if status != "COMPLETED":
        print(f"Error: Job did not complete. Final status: {status}", file=sys.stderr)
        sys.exit(1)

    # Get the final job result
    final_job = client.extract.get(job_id)

    # Prepare the output data
    extract_result = final_job.extract_result

    # Build the result JSON structure
    result = {
        "data": {},
        "extract_metadata": {},
    }

    # Extract the data fields
    if extract_result and extract_result.data:
        result["data"] = extract_result.data

    # Extract the metadata including citations
    if extract_result and extract_result.extract_metadata:
        result["extract_metadata"] = extract_result.extract_metadata.model_dump(mode="json")

    # Write the result JSON
    result_path = "/home/user/llamacloud-task/result.json"
    with open(result_path, "w") as f:
        json.dump(result, f, indent=2)

    print(f"Results written to {result_path}")

    # Write the log file
    log_path = "/home/user/llamacloud-task/output.log"
    with open(log_path, "w") as f:
        f.write(f"Extract job: {job_id}\n")
        f.write(f"Status: {status}\n")

    print(f"Log written to {log_path}")
    print("Extraction completed successfully!")

    sys.exit(0)


if __name__ == "__main__":
    main()

File: jobs/2026-05-27__14-51-17/extract_invoice_fields_ts__i5wBJ3T/artifacts/user/project/extract-invoice.ts
import { z } from 'zod';
import { LlamaCloud } from '@llamaindex/llama-cloud';

/**
 * Zod schema for invoice extraction
 */
const invoiceSchema = z.object({
  invoice_number: z.string(),
  invoice_date: z.string(),
  vendor_name: z.string(),
  total_amount: z.number(),
  line_items: z.array(
    z.object({
      description: z.string(),
      quantity: z.number(),
      unit_price: z.number(),
      total: z.number(),
    })
  ),
});

/**
 * Main extraction function
 */
async function extractInvoice() {
  console.log('Initializing LlamaCloud client...');
  const client = new LlamaCloud({
    apiKey: process.env.LLAMA_CLOUD_API_KEY,
  });

  console.log('Uploading invoice.pdf...');
  const file = await client.files.uploadFile({
    fileInput: {
      fileName: 'invoice.pdf',
      filePath: '/home/user/project/invoice.pdf',
    },
    purpose: 'extract',
  });

  console.log(`File uploaded with ID: ${file.id}`);

  console.log('Creating Extract v2 job with agentic tier...');
  const job = await client.extract.create({
    fileInput: {
      fileId: file.id,
    },
    configuration: {
      dataSchema: invoiceSchema.toJSONSchema(),
      extractionTarget: 'per_doc',
      tier: 'agentic',
    },
  });

  console.log(`Extract job created with ID: ${job.id}`);
  console.log('Polling for job completion...');

  let extractResult: any = null;
  const maxPollingAttempts = 60; // 5 minutes with 5-second intervals
  const pollingInterval = 5000; // 5 seconds

  for (let attempt = 0; attempt < maxPollingAttempts; attempt++) {
    const status = await client.extract.retrieve(job.id);

    console.log(`Attempt ${attempt + 1}/${maxPollingAttempts}: Status = ${status.status}`);

    if (status.status === 'COMPLETED') {
      console.log('Job completed successfully!');
      extractResult = status.extractResult;
      break;
    } else if (status.status === 'FAILED') {
      throw new Error(`Extract job failed: ${status.error || 'Unknown error'}`);
    } else if (status.status === 'CANCELLED') {
      throw new Error('Extract job was cancelled');
    }

    // Wait before polling again
    await new Promise((resolve) => setTimeout(resolve, pollingInterval));
  }

  if (extractResult === null) {
    throw new Error('Job did not complete within the timeout period');
  }

  console.log('Writing results to output.json...');
  const fs = await import('fs');
  fs.writeFileSync(
    '/home/user/project/output.json',
    JSON.stringify(extractResult, null, 2),
    'utf-8'
  );

  console.log('Appending summary to output.log...');
  const summary = `Extracted Invoice: ${extractResult.invoice_number} | Vendor: ${extractResult.vendor_name} | Total: ${extractResult.total_amount}\n`;
  fs.appendFileSync('/home/user/project/output.log', summary, 'utf-8');

  console.log('Extraction completed successfully!');
  console.log(`Invoice Number: ${extractResult.invoice_number}`);
  console.log(`Vendor: ${extractResult.vendor_name}`);
  console.log(`Total Amount: ${extractResult.total_amount}`);
  console.log(`Line Items: ${extractResult.line_items?.length || 0}`);
}

// Run the extraction
extractInvoice().catch((error) => {
  console.error('Error during extraction:', error);
  process.exit(1);
});